x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/jwt/JwtSignatureValidator.java

```diff
@@ -274,17 +276,19 @@ public void validate(String tokenPrincipal, SignedJWT signedJWT, ActionListener<
                 validateSignature(signedJWT, jwksAlgs.jwks());
                 listener.onResponse(null);
             } catch (Exception primaryException) {
-                logger.debug(
-                    () -> org.elasticsearch.core.Strings.format(
-                        "Signature verification failed for JWT [%s] reloading JWKSet (was: #[%s] JWKs, #[%s] algs, sha256=[%s])",
-                        tokenPrincipal,
-                        jwksAlgs.jwks().size(),
-                        jwksAlgs.algs().size(),
-                        MessageDigests.toHexString(contentAndJwksAlgs.sha256())
-                    ),
-                    primaryException
+                String message = org.elasticsearch.core.Strings.format(
+                    "Signature verification failed for JWT token [%s] against JWK set sha256=[%s].",
+                    tokenPrincipal,
+                    MessageDigests.toHexString(contentAndJwksAlgs.sha256())
                 );
 
+                if (logger.isTraceEnabled()) {
+                    logger.trace(message, primaryException);
+                } else {
+                    logger.debug(message + " Cause: " + primaryException.getMessage());
+                }
+
+                logger.debug("Attempting to reload JWK set with sha256=[{}]", MessageDigests.toHexString(contentAndJwksAlgs.sha256()));
                 jwkSetLoader.reload(ActionListener.wrap(ignore -> {
                     final JwkSetLoader.ContentAndJwksAlgs maybeUpdatedContentAndJwksAlgs = jwkSetLoader.getContentAndJwksAlgs();
```

Comment:
GitHub does not let me comment below code. But I think there is a minor theoretical issue: The debug logging calls `jwkSetLoader.getContentAndJwksAlgs()` again to print the sha256. But it is not always guaranteed that this call will return the same result as `maybeUpdatedContentAndJwksAlgs` here since the `jwtSetLoader` can in theory completes another reload round.

Reply:
good catch, fixed. hopefully the logging at [debug](https://github.com/elastic/elasticsearch/pull/98758#discussion_r1304888635) level will help to catch if there are any other concurrent issues here.